runtime/codefrag.c

```diff
@@ -0,0 +1,157 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cambium, INRIA Paris                  */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+/* A table of all code fragments (main program and dynlinked modules) */
+
+#include <string.h>
+#include <stddef.h>
+#include "caml/codefrag.h"
+#include "caml/misc.h"
+#include "caml/md5.h"
+#include "caml/memory.h"
+#include "caml/skiplist.h"
+
+static struct skiplist code_fragments_by_pc = SKIPLIST_STATIC_INITIALIZER;
+
+static struct skiplist code_fragments_by_num = SKIPLIST_STATIC_INITIALIZER;
+
+static int code_fragments_counter = 0;
+
+static struct skiplist code_fragments_by_digest = SKIPLIST_STATIC_INITIALIZER;
+
+static struct code_fragment * code_fragments_recently_added = NULL;
+
+int caml_register_code_fragment(char * start, char * end,
+                                enum digest_status digest_kind,
+                                unsigned char * opt_digest)
+{
+  struct code_fragment * cf = caml_stat_alloc(sizeof(struct code_fragment));
+
+  cf->code_start = start;
+  cf->code_end = end;
+  switch (digest_kind) {
+  case DIGEST_LATER:
+    break;
+  case DIGEST_NOW:
+    caml_md5_block(cf->digest, cf->code_start, cf->code_end - cf->code_start);
+    digest_kind = DIGEST_PROVIDED;
+    break;
+  case DIGEST_PROVIDED:
+    memcpy(cf->digest, opt_digest, 16);
+    break;
+  case DIGEST_IGNORE:
+    break;
+  }
+  cf->digest_status = digest_kind;
+  cf->fragnum = code_fragments_counter++;
+  caml_skiplist_insert(&code_fragments_by_pc,
+                       (uintnat) start, (uintnat) cf);
+  caml_skiplist_insert(&code_fragments_by_num,
+                       cf->fragnum, (uintnat) cf);
```

Comment:
```suggestion
                       (uintnat) cf->fragnum, (uintnat) cf);
```
I think it's better to have an explicit rather than implicit cast.

Reply:
All right.  See commit dc6c33fd7